runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1204,36 +1206,38 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
-               //
-               if (storeValueBaseNode == NULL || getValueNumber(storeValueBaseNode) != getValueNumber(arrayRefNode))
+               if (!owningMethodDoesNotContainStoreChecks(this, node))
                   {
-                  // If storing to an array whose component type is or might be a primitive value
-                  // type and the value that's being assigned is or might be null, both a run-time
-                  // NULLCHK of the value is required (guarded by a check of whether the
-                  // component type is a value type) and an ArrayStoreCHK are required;
-                  // otherwise, only the ArrayStoreCHK is required.
+                  TR::Node *storeValueBaseNode = getStoreValueBaseNode(storeValueNode, symRefTab);
+
+                  // If the value was loaded from the same array that is the target of the store, no need
+                  // to perform ArrayStoreCHK.
                   //
-                  bool mustFail = false;
-                  if (isArrayStoreCheckNeeded(arrayRefNode, storeValueNode, mustFail, storeClassForCheck, componentClassForCheck))
+                  if (storeValueBaseNode == NULL || getValueNumber(storeValueBaseNode) != getValueNumber(arrayRefNode))
                      {
-                     flagsForTransform.set(ValueTypesHelperCallTransform::RequiresStoreCheck);
+                     // If storing to an array whose component type is or might be a primitive value
+                     // type and the value that's being assigned is or might be null, both a run-time
+                     // NULLCHK of the value is required (guarded by a check of whether the
+                     // component type is a value type) and an ArrayStoreCHK are required;
+                     // otherwise, only the ArrayStoreCHK is required.
+                     //
+                     bool mustFail = false;
+                     if (isArrayStoreCheckNeeded(arrayRefNode, storeValueNode, mustFail, storeClassForCheck, componentClassForCheck))
+                        {
+                        flagsForTransform.set(ValueTypesHelperCallTransform::RequiresStoreCheck);
+                        }
                      }
+                  }
 
-                  if (canTransformUnflattenedArrayElementLoadStoreUseTypeHint && (!storeValueConstraint || !storeValueConstraint->isNonNullObject()))
-                     {
-                     flagsForTransform.set(ValueTypesHelperCallTransform::RequiresNullValueCheck);
-                     }
+               //TODO: Require the <nonNullableArrayNullStoreCheck> non-helper if !canSkipNonNullableArrayNullValueChecks(...)
+               if (canTransformUnflattenedArrayElementLoadStoreUseTypeHint &&
+                  (!storeValueConstraint || !storeValueConstraint->isNonNullObject()))
+                  {
+                  flagsForTransform.set(ValueTypesHelperCallTransform::RequiresNullValueCheck);
                   }
```

Comment:
As I mentioned above, the way this code was organized before, Value Propagation was able to avoid the call to `<nonNullableArrayNullStoreCheck>` if the value being stored to the array had been loaded from the same array.  That optimization should still be possible, even if the concept of `owningMethodDoesNotContainNonNullableArrayNullValueChecks` is separated from `owningMethodDoesNotContainStoreChecks`.

Reply:
Addressed in 156b344